### ISSUE TITLE
Feature Request: Memory-Efficient Differentiable Rendering

### ISSUE THREAD

**Issue by ashishv-git** [optiland/optiland#689]
### Checklist

- [X] I have searched the [existing issues](https://github.com/HarrisonKramer/optiland/issues) and [discussions](https://github.com/HarrisonKramer/optiland/discussions) for a similar question or feature request.
- [X] I have read the [documentation](https://optiland.readthedocs.io/en/latest/) and tried to find an answer there.
- [X] I am using the **latest version** of Optiland.
- [X] I have included **all necessary context**.

### Feature Request

**Is your feature request related to a problem? Please describe.**
This proposal is a follow-up to the discussion in Issue #335 by @manuelFragata regarding efficiency in Differentiable Ray Tracing (DRT).

In #335, there was a discussion regarding the "dO: A differentiable engine for Deep Lens design
 of computational imaging systems" (https://github.com/vccimaging/DiffOptics) paper, specifically focusing on ray-surface intersection.

PR #554 addresses the efficiency of the Newton-Raphson root solver via implicit differentiation. The *dO* paper proposed another memory-saving approach that enables tracing a large number of sampling rays by splitting backpropagation. Without this approach, the global memory usage scales rapidly with the number of rays. Consider an end-to-end design of a computational imaging system where the computational graph spans from the optical system parameters, through a rendered image, and into a neural network (e.g., for perceptual loss evaluation).  When tracing millions of rays for dense image rendering (e.g., rendering an image and passing it through a CNN for perceptual loss), the standard PyTorch Autograd engine stores the entire computational graph in memory. Consequently, without a mechanism to split the backpropagation, the global memory usage scales rapidly with the number of rays.

I wrote a profiling script to measure this behavior. Please see the graph below:

<img width="3000" height="1200" alt="Image" src="https://github.com/user-attachments/assets/0c67d95a-749b-4c60-9d17-42b474ef55a1" />

*(Note: The profiling script used to generate this plot is included:)*

[profile_standard_bp.py](https://github.com/user-attachments/files/30154688/profile_standard_bp.py)

**Describe the solution you'd like**
The *dO* paper uses an approach termed **Adjoint Backpropagation** (Section II-D of the paper).

It uses the "separability" of the computational graph that allows decoupling of the optical and non-optical parameters' gradients. This approach is implemented in three stages:
1. First, a forward render (without AD graph construction).
2. Then, forward and backward passes on the metric function (over the neural network parameters) with respect to the rendered image.
3. Finally, forward and backward passes on the optical system.

This "separability property" allows us to compute the Vector-Jacobian Products in small chunks of rays which allows the memory footprint to remain relatively constant.

**Describe alternatives you've considered**
N/A

**Additional context**
I would be happy to work on this feature if it aligns with the goals of the project.

**Comment by HarrisonKramer** [comment]
@ashishv-git

Thanks for the writeup, and especially for the profiling script and plot. That memory scaling is a real limitation for end-to-end computational imaging work in Optiland, so this is a direction I am happy to support. Assigning it to you.

One scoping thought. The most valuable piece here is the chunked VJP accumulation itself, independent of any particular renderer: forward under `no_grad`, then re-run in ray chunks accumulating VJPs against the incoming adjoint. That applies to anything reducing additively over rays, not only image rendering, so it pays off across the library. I would rather see it land as a clean standalone primitive than wired into a single rendering path. If you see it differently once you are into the code, say so.

A few project conventions that will shape the design:

* The public API is frozen and we prefer atomic PRs, so please add this as new surface area rather than changing existing signatures.
* All numerical code goes through `optiland.backend as be`. This feature is inherently torch-only, so guard it explicitly and make sure the numpy path stays untouched and fails clearly.
* This is a gradient-accumulation strategy, not a tracing concern, so it should not be bolted into the tracer. `optiland/ml/`, alongside `OpticalSystemModule`, is the natural home IMO.
* Tests are the gate: gradient equivalence against plain non-chunked autograd, and invariance of the result to chunk size.

Two things that can silently yield wrong gradients, worth designing against from the start. Chunking is only valid when the reduction over rays is genuinely additive, so it breaks for reductions like max and for anything normalized by a quantity derived from all rays, such as a centroid reference. And the chunked pass has to reproduce the no-grad pass exactly, so anything stochastic in ray generation or scattering needs care.

Worth noting this composes with #554 rather than overlapping it: implicit differentiation bounds graph depth per ray, chunking bounds width across rays.

Feel free to open a draft PR early and I will review as it develops.

Regards,
Kramer

**Comment by ashishv-git** [comment]
Hi Kramer,

Thanks a lot for the detailed feedback.

I appreciate the notes on scoping. I will build the chunked VJP accumulation as a standalone primitive in `optiland/ml/` rather than tying it to a specific renderer.

For the implementation, I will make sure to:
* Explicitly check that the active backend is Torch and fail clearly if it is NumPy.
* Add tests for gradient equivalence against standard non-chunked autograd, as well as chunk size invariance.
* Restrict the feature to additive reductions, and manage seeds for stochastic ray generation and scattering so the chunked pass correctly reproduces the initial `no_grad` pass.

I will get started on this and open a draft PR soon so you can review the approach as it develops.

Regards,
Ashish
